### DISCUSSION TITLE
How can I customize a port for a specific triplet in vcpkg?

### DISCUSSION THREAD

**Discussion by SeanTolstoyevski** [microsoft/vcpkg#41076] [Q&A]
Hello,

For the Poco port, I need to make some changes to the Community triplets x64-mingw and x64-mingw-static.
In summary, if Poco is compiled with MinGW, some definitions are required.
One of them is the `-DPOCO_COMPILER_MINGW` flag.

This flag is not set with the current port and we see the following error:

```bash
D:\projs\vcpkg>vcpkg.exe install poco:x64-mingw-static
Computing installation plan...
The following packages will be built and installed:
    poco:x64-mingw-static@1.13.3
Detecting compiler hash for triplet x64-mingw-static...
Compiler found: D:/Apps/msys64/clang64/bin/c++.exe
Restored 0 package(s) from C:\Users\e\AppData\Local\vcpkg\archives in 128 us. Use --debug to see more details.
Installing 1/1 poco:x64-mingw-static@1.13.3...
Building poco:x64-mingw-static@1.13.3...
D:\projs\vcpkg\triplets\community\x64-mingw-static.cmake: info: loaded community triplet from here. Community triplets are not built in the curated registry and are thus less likely to succeed.
-- VCPKG_TARGET_TRIPLET:
-- Using cached pocoproject-poco-poco-1.13.3-release.tar.gz.
-- Cleaning sources at D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/projs/vcpkg/downloads/pocoproject-poco-poco-1.13.3-release.tar.gz
-- Applying patch 0001-static-pcre.patch
-- Applying patch 0002-arm64-pcre.patch
-- Applying patch 0003-fix-dependency.patch
-- Applying patch 0004-fix-feature-sqlite3.patch
-- Applying patch 0005-fix-error-c3861.patch
-- Applying patch 0007-find-pcre2.patch
-- Using source at D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean
-- Configuring x64-mingw-static
-- Building x64-mingw-static-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: "C:/Program Files/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v -j13
    Working Directory: D:/projs/vcpkg/buildtrees/poco/x64-mingw-static-dbg
    See logs for more information:
      D:\projs\vcpkg\buildtrees\poco\install-x64-mingw-static-dbg-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/poco/portfile.cmake:93 (vcpkg_cmake_install)
  scripts/ports.cmake:192 (include)


error: building poco:x64-mingw-static failed with: BUILD_FAILED
See https://learn.microsoft.com/vcpkg/troubleshoot/build-failures?WT.mc_id=vcpkg_inproduct_cli for more information.
Elapsed time to handle poco:x64-mingw-static: 46 s
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+poco
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?title=[poco]+Build+error+on+x64-mingw-static&body=Copy+issue+body+from+D%3A%2Fprojs%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md
```

And __output log__ `D:\projs\vcpkg\buildtrees\poco\install-x64-mingw-static-dbg-out.log`:

```
...............

[60/626] D:\Apps\msys64\clang64\bin\c++.exe --target=x86_64-windows-gnu -DMINGW32 -DODBCVER=0x0300 -DPOCO_ENABLE_CPP11 -DPOCO_ENABLE_CPP14 -DPOCO_NO_AUTOMATIC_LIBS -DPOCO_OS_FAMILY_WINDOWS -DPOCO_STATIC -DPOCO_THREAD_STACK_SIZE -DPOCO_UNBUNDLED -DPOCO_WIN32_UTF8 -DUNICODE -DWC_NO_BEST_FIT_CHARS=0x400 -DWINVER=0x500 -D_DEBUG -D_UNICODE -D_WIN32 -D_pcre2_OP_lengths_8=_poco_pcre2_OP_lengths_8 -D_pcre2_callout_end_delims_8=_poco_pcre2_callout_end_delims_8 -D_pcre2_callout_start_delims_8=_poco_pcre2_callout_start_delims_8 -D_pcre2_hspace_list_8=_poco_pcre2_hspace_list_8 -D_pcre2_ucp_gbtable_8=_poco_pcre2_ucp_gbtable_8 -D_pcre2_ucp_gentype_8=_poco_pcre2_ucp_gentype_8 -D_pcre2_utf8_table1=_poco_pcre2_utf8_table1 -D_pcre2_utf8_table1_size=_poco_pcre2_utf8_table1_size -D_pcre2_utf8_table2=_poco_pcre2_utf8_table2 -D_pcre2_utf8_table3=_poco_pcre2_utf8_table3 -D_pcre2_utf8_table4=_poco_pcre2_utf8_table4 -D_pcre2_utt_8=_poco_pcre2_utt_8 -D_pcre2_utt_names_8=_poco_pcre2_utt_names_8 -D_pcre2_utt_size_8=_poco_pcre2_utt_size_8 -D_pcre2_vspace_list_8=_poco_pcre2_vspace_list_8 -ID:/projs/vcpkg/buildtrees/poco/x64-mingw-static-dbg/Foundation -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/src -isystem D:/projs/vcpkg/installed/x64-mingw-static/include -g -std=gnu++17 -MD -MT Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.obj -MF Foundation\CMakeFiles\Foundation.dir\src\Thread.cpp.obj.d -o Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.obj -c D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/src/Thread.cpp
FAILED: Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.obj 
D:\Apps\msys64\clang64\bin\c++.exe --target=x86_64-windows-gnu -DMINGW32 -DODBCVER=0x0300 -DPOCO_ENABLE_CPP11 -DPOCO_ENABLE_CPP14 -DPOCO_NO_AUTOMATIC_LIBS -DPOCO_OS_FAMILY_WINDOWS -DPOCO_STATIC -DPOCO_THREAD_STACK_SIZE -DPOCO_UNBUNDLED -DPOCO_WIN32_UTF8 -DUNICODE -DWC_NO_BEST_FIT_CHARS=0x400 -DWINVER=0x500 -D_DEBUG -D_UNICODE -D_WIN32 -D_pcre2_OP_lengths_8=_poco_pcre2_OP_lengths_8 -D_pcre2_callout_end_delims_8=_poco_pcre2_callout_end_delims_8 -D_pcre2_callout_start_delims_8=_poco_pcre2_callout_start_delims_8 -D_pcre2_hspace_list_8=_poco_pcre2_hspace_list_8 -D_pcre2_ucp_gbtable_8=_poco_pcre2_ucp_gbtable_8 -D_pcre2_ucp_gentype_8=_poco_pcre2_ucp_gentype_8 -D_pcre2_utf8_table1=_poco_pcre2_utf8_table1 -D_pcre2_utf8_table1_size=_poco_pcre2_utf8_table1_size -D_pcre2_utf8_table2=_poco_pcre2_utf8_table2 -D_pcre2_utf8_table3=_poco_pcre2_utf8_table3 -D_pcre2_utf8_table4=_poco_pcre2_utf8_table4 -D_pcre2_utt_8=_poco_pcre2_utt_8 -D_pcre2_utt_names_8=_poco_pcre2_utt_names_8 -D_pcre2_utt_size_8=_poco_pcre2_utt_size_8 -D_pcre2_vspace_list_8=_poco_pcre2_vspace_list_8 -ID:/projs/vcpkg/buildtrees/poco/x64-mingw-static-dbg/Foundation -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/src -isystem D:/projs/vcpkg/installed/x64-mingw-static/include -g -std=gnu++17 -MD -MT Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.obj -MF Foundation\CMakeFiles\Foundation.dir\src\Thread.cpp.obj.d -o Foundation/CMakeFiles/Foundation.dir/src/Thread.cpp.obj -c D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/src/Thread.cpp
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/src/Thread.cpp:24:
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/src/Thread_WIN32.cpp:49:3: error: use of undeclared identifier '__try'
   49 |                 __try
      |                 ^
1 error generated.

...............
```

The reason for the error is that the `__try` statement can only be __used with MSVC__. And these lines are compiled with __conditional compilation (ifdef ifndef etc.)__, but Poco does not detect this due to the missing flag when compiling.

I also want to customize this Poco port for these 2 triplets.

To try it I made the following additions to the `ports/Poco/portfile.cmake` file, but it didn't work:

```cmake
message(STATUS "VCPKG_TARGET_TRIPLET: ${VCPKG_TARGET_TRIPLET}") # only debugging
if (VCPKG_TARGET_TRIPLET STREQUAL "x64-mingw-static") # not handled because VCPKG_TARGET_TRIPLET returned  empty value.
    message(STATUS "triggered")
    list(APPEND VCPKG_CXX_FLAGS "-DPOCO_COMPILER_MINGW=1")
    list(APPEND VCPKG_C_FLAGS "-DPOCO_COMPILER_MINGW=1")
endif()

# continue original code
```

I also read the available ports, but I could not understand how such specific things are done or what vcpkg's approach is to this issue.

Can you give me an idea about this?

**Comment by dg0yt** [comment]
> In summary, if Poco is compiled with MinGW, some definitions are required. One of them is the `-DPOCO_COMPILER_MINGW` flag.
> 
> This flag is not set with the current port

I don't think this statement is right. POCO takes care of it:
https://github.com/pocoproject/poco/blob/cd06553d89aee4bf7b5c820199121cd25d5bce94/Foundation/include/Poco/Platform.h#L259-L260

> ~~~cmake
>     list(APPEND VCPKG_CXX_FLAGS "-DPOCO_COMPILER_MINGW=1")
>     list(APPEND VCPKG_C_FLAGS "-DPOCO_COMPILER_MINGW=1")
> ~~~

Note that these variables, like `CMAKE_<lang>_FLAGS`, are command line fragments. You would need to use:
~~~cmake
    string(APPEND VCPKG_CXX_FLAGS " -DPOCO_COMPILER_MINGW")
    string(APPEND VCPKG_C_FLAGS " -DPOCO_COMPILER_MINGW")
~~~
(Note the space!)

**Reply by SeanTolstoyevski** [reply]
Thank you for your answer. 
Especially your last cmake lines made my job easier.
I changed portfile.cmake as follows and the previous error disappeared but I continue to get other errors. 

I will research these and try to find solutions.

```cmake
# VCPKG_TARGET_IS_MINGW
if (VCPKG_TARGET_IS_MINGW)
    message(STATUS "triggered")
    string(APPEND VCPKG_CXX_FLAGS " -DPOCO_COMPILER_MINGW")
    string(APPEND VCPKG_C_FLAGS " -DPOCO_COMPILER_MINGW")
endif()

# original code
```


vcpkg output:  
```bash
D:\projs\vcpkg>vcpkg install  poco:x64-mingw-static
Computing installation plan...
The following packages will be built and installed:
    poco:x64-mingw-static@1.13.3
Detecting compiler hash for triplet x64-mingw-static...
Compiler found: D:/Apps/msys64/clang64/bin/c++.exe
Restored 0 package(s) from C:\Users\e\AppData\Local\vcpkg\archives in 91.7 us. Use --debug to see more details.
Installing 1/1 poco:x64-mingw-static@1.13.3...
Building poco:x64-mingw-static@1.13.3...
D:\projs\vcpkg\triplets\community\x64-mingw-static.cmake: info: loaded community triplet from here. Community triplets are not built in the curated registry and are thus less likely to succeed.
-- triggered
-- Using cached pocoproject-poco-poco-1.13.3-release.tar.gz.
-- Cleaning sources at D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/projs/vcpkg/downloads/pocoproject-poco-poco-1.13.3-release.tar.gz
-- Applying patch 0001-static-pcre.patch
-- Applying patch 0002-arm64-pcre.patch
-- Applying patch 0003-fix-dependency.patch
-- Applying patch 0004-fix-feature-sqlite3.patch
-- Applying patch 0005-fix-error-c3861.patch
-- Applying patch 0007-find-pcre2.patch
-- Using source at D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean
-- Configuring x64-mingw-static
-- Building x64-mingw-static-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: "C:/Program Files/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v -j13
    Working Directory: D:/projs/vcpkg/buildtrees/poco/x64-mingw-static-dbg
    See logs for more information:
      D:\projs\vcpkg\buildtrees\poco\install-x64-mingw-static-dbg-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/poco/portfile.cmake:91 (vcpkg_cmake_install)
  scripts/ports.cmake:192 (include)


error: building poco:x64-mingw-static failed with: BUILD_FAILED
See https://learn.microsoft.com/vcpkg/troubleshoot/build-failures?WT.mc_id=vcpkg_inproduct_cli for more information.
Elapsed time to handle poco:x64-mingw-static: 2 min
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+poco
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?title=[poco]+Build+error+on+x64-mingw-static&body=Copy+issue+body+from+D%3A%2Fprojs%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md`


output (only error lines): `D:/projs/vcpkg/buildtrees/poco/install-x64-mingw-static-dbg-out.log` :
```
[216/626] D:\Apps\msys64\clang64\bin\c++.exe --target=x86_64-windows-gnu -DMINGW32 -DODBCVER=0x0300 -DPOCO_ENABLE_CPP11 -DPOCO_ENABLE_CPP14 -DPOCO_NO_AUTOMATIC_LIBS -DPOCO_OS_FAMILY_WINDOWS -DPOCO_STATIC -DPOCO_THREAD_STACK_SIZE -DPOCO_UNBUNDLED -DPOCO_WIN32_UTF8 -DSQLParser_EXPORTS -DTHREADSAFE -DUNICODE -DWC_NO_BEST_FIT_CHARS=0x400 -DWINVER=0x500 -D_DEBUG -D_UNICODE -D_WIN32 -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/include -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/include -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/src -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include -isystem D:/projs/vcpkg/installed/x64-mingw-static/include -DPOCO_COMPILER_MINGW -g -std=gnu++17 -MD -MT Data/ODBC/CMakeFiles/DataODBC.dir/src/ODBCMetaColumn.cpp.obj -MF Data\ODBC\CMakeFiles\DataODBC.dir\src\ODBCMetaColumn.cpp.obj.d -o Data/ODBC/CMakeFiles/DataODBC.dir/src/ODBCMetaColumn.cpp.obj -c D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/src/ODBCMetaColumn.cpp
FAILED: Data/ODBC/CMakeFiles/DataODBC.dir/src/ODBCMetaColumn.cpp.obj 
D:\Apps\msys64\clang64\bin\c++.exe --target=x86_64-windows-gnu -DMINGW32 -DODBCVER=0x0300 -DPOCO_ENABLE_CPP11 -DPOCO_ENABLE_CPP14 -DPOCO_NO_AUTOMATIC_LIBS -DPOCO_OS_FAMILY_WINDOWS -DPOCO_STATIC -DPOCO_THREAD_STACK_SIZE -DPOCO_UNBUNDLED -DPOCO_WIN32_UTF8 -DSQLParser_EXPORTS -DTHREADSAFE -DUNICODE -DWC_NO_BEST_FIT_CHARS=0x400 -DWINVER=0x500 -D_DEBUG -D_UNICODE -D_WIN32 -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/include -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/include -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/src -ID:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include -isystem D:/projs/vcpkg/installed/x64-mingw-static/include -DPOCO_COMPILER_MINGW -g -std=gnu++17 -MD -MT Data/ODBC/CMakeFiles/DataODBC.dir/src/ODBCMetaColumn.cpp.obj -MF Data\ODBC\CMakeFiles\DataODBC.dir\src\ODBCMetaColumn.cpp.obj.d -o Data/ODBC/CMakeFiles/DataODBC.dir/src/ODBCMetaColumn.cpp.obj -c D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/src/ODBCMetaColumn.cpp
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/src/ODBCMetaColumn.cpp:169:7: error: use of undeclared identifier 'SQL_GUID'
  169 |         case SQL_GUID:
      |              ^
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/src/ODBCMetaColumn.cpp:15:
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/include/Poco/Data/ODBC/ODBCMetaColumn.h:22:
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/include/Poco/Data/ODBC/Error.h:22:
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/include/Poco/Data/ODBC/Utility.h:22:
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Data/ODBC/include/Poco/Data/ODBC/TypeInfo.h:22:
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/NamedTuple.h:24:
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/DynamicAny.h:21:
In file included from D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/Var.h:26:
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:424:15: warning: implicit conversion from 'type' (aka 'int') to 'float' changes value from 2147483647 to 2147483648 [-Wimplicit-const-int-float-conversion]
  424 |                         if (from > std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:327:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkUpperLimit<float, int, true>' requested here
  327 |                 checkUpperLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2189:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertToSmaller<float, int>' requested here
 2189 |                 convertToSmaller(_val, val);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:440:15: warning: implicit conversion from 'type' (aka 'int') to 'float' changes value from -2147483647 to -2147483648 [-Wimplicit-const-int-float-conversion]
  440 |                         if (from < -std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:328:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkLowerLimit<float, int, true>' requested here
  328 |                 checkLowerLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2189:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertToSmaller<float, int>' requested here
 2189 |                 convertToSmaller(_val, val);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:424:15: warning: implicit conversion from 'type' (aka 'long long') to 'float' changes value from 9223372036854775807 to 9223372036854775808 [-Wimplicit-const-int-float-conversion]
  424 |                         if (from > std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:327:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkUpperLimit<float, long long, true>' requested here
  327 |                 checkUpperLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2194:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertToSmaller<float, long long>' requested here
 2194 |                 convertToSmaller(_val, val);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:440:15: warning: implicit conversion from 'type' (aka 'long long') to 'float' changes value from -9223372036854775807 to -9223372036854775808 [-Wimplicit-const-int-float-conversion]
  440 |                         if (from < -std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:328:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkLowerLimit<float, long long, true>' requested here
  328 |                 checkLowerLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2194:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertToSmaller<float, long long>' requested here
 2194 |                 convertToSmaller(_val, val);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:424:15: warning: implicit conversion from 'type' (aka 'unsigned int') to 'float' changes value from 4294967295 to 4294967296 [-Wimplicit-const-int-float-conversion]
  424 |                         if (from > std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:383:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkUpperLimit<float, unsigned int, true>' requested here
  383 |                 checkUpperLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2209:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertSignedFloatToUnsigned<float, unsigned int, true>' requested here
 2209 |                 convertSignedFloatToUnsigned(_val, val);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:424:15: warning: implicit conversion from 'type' (aka 'unsigned long long') to 'float' changes value from 18446744073709551615 to 18446744073709551616 [-Wimplicit-const-int-float-conversion]
  424 |                         if (from > std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:383:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkUpperLimit<float, unsigned long long, true>' requested here
  383 |                 checkUpperLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2214:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertSignedFloatToUnsigned<float, unsigned long long, true>' requested here
 2214 |                 convertSignedFloatToUnsigned(_val, val);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:424:15: warning: implicit conversion from 'type' (aka 'long long') to 'double' changes value from 9223372036854775807 to 9223372036854775808 [-Wimplicit-const-int-float-conversion]
  424 |                         if (from > std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:327:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkUpperLimit<double, long long, true>' requested here
  327 |                 checkUpperLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2347:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertToSmaller<double, long long>' requested here
 2347 |                 convertToSmaller(_val, val);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:440:15: warning: implicit conversion from 'type' (aka 'long long') to 'double' changes value from -9223372036854775807 to -9223372036854775808 [-Wimplicit-const-int-float-conversion]
  440 |                         if (from < -std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:328:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkLowerLimit<double, long long, true>' requested here
  328 |                 checkLowerLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2347:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertToSmaller<double, long long>' requested here
 2347 |                 convertToSmaller(_val, val);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:424:15: warning: implicit conversion from 'type' (aka 'unsigned long long') to 'double' changes value from 18446744073709551615 to 18446744073709551616 [-Wimplicit-const-int-float-conversion]
  424 |                         if (from > std::numeric_limits<T>::max())
      |                                  ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:383:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::checkUpperLimit<double, unsigned long long, true>' requested here
  383 |                 checkUpperLimit<F,T>(from);
      |                 ^
D:/projs/vcpkg/buildtrees/poco/src/.3-release-bdefce8352.clean/Foundation/include/Poco/Dynamic/VarHolder.h:2367:3: note: in instantiation of function template specialization 'Poco::Dynamic::VarHolder::convertSignedFloatToUnsigned<double, unsigned long long, true>' requested here
 2367 |                 convertSignedFloatToUnsigned(_val, val);
      |                 ^
9 warnings and 1 error generated.
```

***

If you know about Poco, do you have a chance to tell me why this is happening? I read the code but I couldn't find the problem.

**Reply by dg0yt** [reply]
I guess there is not enough testing of Poco with MinGW. I assume you use a fairly up-to-date installation of MSYS2, so the MinGW version shouldn't be an issue.

MSYS2 also has Poco packages for 1.13.0. Their patches might also give a clue:  
https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-poco  
in particular  
https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-poco/001-fix-build-on-mingw.patch
